lib/streamlit/navigation/page.py

```diff
@@ -0,0 +1,168 @@
+
+    Parameters
+    ----------
+
+    page: str or Path or callable
+        The path to the script file or a callable that defines the page.
+        The path must be relative to the main script.
+
+    title: str or None
+        The title of the page. If None, the title will be inferred from the
+        page path or callable name.
+
+    icon: str or None
+        An optional emoji or icon to display next to the alert. If ``icon``
+        is ``None`` (default), no icon is displayed. If ``icon`` is a
+        string, the following options are valid:
+
+        * A single-character emoji. For example, you can set ``icon="🚨"``
+            or ``icon="🔥"``. Emoji short codes are not supported.
+
+        * An icon from the Material Symbols library (outlined style) in the
+            format ``":material/icon_name:"`` where "icon_name" is the name
+            of the icon in snake case.
+
+            For example, ``icon=":material/thumb_up:"`` will display the
+            Thumb Up icon. Find additional icons in the `Material Symbols \
+            <https://fonts.google.com/icons?icon.set=Material+Symbols&icon.style=Outlined>`_
+            font library.
+
+    default: bool
+        Whether this page is the default page to be shown when the app is
+        loaded. Only one page can be marked default. If no default page is
+        provided, the first page will be the default page.
+
+    Example
+    -------
+    >>> import streamlit as st
+    >>>
+    >>> def page2():
+    >>>     st.title("Second page")
+    >>> pg = st.navigation([
+    >>>	    st.Page("page1.py", title="First page", icon="🔥"),
+    >>>	    st.Page(page2, title="Second page", icon=":material/favorite:"),
+    >>> ])
+    >>> pg.run()
+    """
+    return StreamlitPage(page, title=title, icon=icon, default=default)
+
+
+class StreamlitPage:
+    def __init__(
+        self,
+        page: str | Path | Callable[[], None],
+        *,
+        title: str | None = None,
+        icon: str | None = None,
+        default: bool = False,
+    ):
+        ctx = get_script_run_ctx()
+        if not ctx:
+            return
+
+        main_path = Path(ctx.pages_manager.main_script_path).parent
+        if isinstance(page, str):
+            page = Path(page)
+        if isinstance(page, Path):
+            page = (main_path / page).resolve()
+
+        inferred_name = ""
+        inferred_icon = ""
+        if isinstance(page, Path):
+            inferred_icon, inferred_name = page_icon_and_name(page)
+        elif title is None:
+            if hasattr(page, "__name__"):
+                inferred_name = str(page.__name__)
+            else:
+                raise StreamlitAPIException(
+                    "Cannot infer page title for Callable. Set the `title=` keyword argument."
+                )
+
+        self._page: Path | Callable[[], None] = page
+        self.title: str = title or inferred_name
+        self.icon: str = icon or inferred_icon
```

Comment:
Should these (and `page_url`) be properties to internal variables, to prevent people from manipulating them after initialization and potentially causing weird behavior?

Reply:
Certainly possible to prevent the idea of changing. I don't see a problem with users changing it, but we can be conservative here as well.